// eslint-plugin-import/lib/rules/newline-after-import.js
'use strict';

var _staticRequire = require('../core/staticRequire');

var _staticRequire2 = _interopRequireDefault(_staticRequire);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

function getLineDifference(node, nextToken) {
  return nextToken.loc.start.line - node.loc.start.line;
} /**
   * @fileoverview Rule to enforce new line after import not followed by another import.
   * @author Radek Benkel
   */

function ensureNoForbiddenKeyword(context, node, tokenToInspect, tokenValue) {
  if (!tokenToInspect) {
    return;
  }

  if (getLineDifference(node, tokenToInspect) === 1 && tokenToInspect.type === 'Keyword' && tokenToInspect.value !== tokenValue) {
    context.report({
      loc: tokenToInspect.loc.start,
      message: 'Expected empty line after ' + tokenValue + ' statement not followed by another ' + tokenValue + '.'
    });
  }
}

module.exports = function (context) {
  return {
    ImportDeclaration: function ImportDeclaration(node) {
      var nextToken = context.getSourceCode(node).getTokenAfter(node);

      ensureNoForbiddenKeyword(context, node, nextToken, 'import');
    },
    CallExpression: function CallExpression(node) {
      if ((0, _staticRequire2.default)(node)) {
        var nextTokens = context.getSourceCode(node).getTokensAfter(node, 2);
        var tokenToInspect = nextTokens.length > 1 && nextTokens[0].type === 'Punctuator' ? nextTokens[1] : nextTokens[0];

        ensureNoForbiddenKeyword(context, node, tokenToInspect, 'require');
      }
    }
  };
};
